### PR TITLE
Variables recursion options

### PR DESCRIPTION
**Pull request by pjazdzewski1990**

### CODE REVIEW

File: src/test/scala/eu/gruchala/scala/language/OptionsSpec.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    def sillyJavaStyleFindById(id: Int) = if(id % 2 == 0) id else null
+    // the method above is dangerous, let's fix that!
+    val existingValue = Option(sillyJavaStyleFindById(42))
+    val missingValue = Option(sillyJavaStyleFindById(1))
+
+    it("allows us to get the value in few different ways, some are risky ... ") {
+      existingValue.get shouldBe 42
+      println(s" missing VALYES $missingValue")
```
Comment (leszekgruchala) [inline src/test/scala/eu/gruchala/scala/language/OptionsSpec.scala]: Please get rid of `println` in specs and use value with matchers.

File: src/test/scala/eu/gruchala/scala/language/OptionsSpec.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      if(!missingValue.isEmpty) {
+        fail("this will not happen")
+      }
+    }
+
+    it("... others are handy ..."){
+      existingValue.getOrElse(0) shouldBe 42
+      missingValue.getOrElse(0) shouldBe 0
+      missingValue.getOrElse({
+        println("Sometimes you want to calculate a proper result")
```
Comment (leszekgruchala) [inline src/test/scala/eu/gruchala/scala/language/OptionsSpec.scala]: As above.

File: src/test/scala/eu/gruchala/scala/language/VariablesSpec.scala [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+      val itsInferredFromFunctions = returnString()
+      itsInferredFromFunctions shouldBe a [String]
+
+      val itsInferredFromFunctionsWithoutType = returnTypesCanBeInferredToo()
+      itsInferredFromFunctionsWithoutType shouldBe a [String]
+    }
+
+    it("can deferred until really needed - we call them lazy") {
+      lazy val imLazy = new Date()
```
Comment (leszekgruchala) [inline src/test/scala/eu/gruchala/scala/language/VariablesSpec.scala:45]: Maybe show that further calls to `lazy val` will return the same value/reference?
`Thread.sleep` in specs doesn't look like a good practice.

Comment (pjazdzewski1990) [inline src/test/scala/eu/gruchala/scala/language/VariablesSpec.scala:45]: `Thread.sleep` isn't an good practice .. everywhere not only in tests ;)

**Pull request merged**